src/framegrab/grabber.py

```diff
@@ -1,253 +1,759 @@
+            if serial_number is None or serial_number == curr_serial_number:
+                camera = pylon.InstantCamera(pylon.TlFactory.GetInstance().CreateDevice(device))
+                camera.Open()
+                logger.info(f"Connected to Basler USB camera with serial number {curr_serial_number}.")
+                break
+        else:
+            raise ValueError(
+                f"Unable to connect to Basler USB camera with serial number: {serial_number}. "
+                "Please verify that the camera is plugged in and that the serial number is correct."
+            )
 
+        # A valid camera has been found, remember the serial_number to prevent
+        # other FrameGrabbers from using it
+        self.camera = camera
+        self.serial_number = curr_serial_number
+        BaslerUSBFrameGrabber.serial_numbers_in_use.add(self.serial_number)
 
-class ImageURLFrameGrabber(FrameGrabber):
-    """grabs the current image at a single URL.
-    NOTE: if image is expected to be refreshed or change with a particular frequency,
-    it is up to the user of the class to call the `grab` method with that frequency
-    """
+    def grab(self) -> np.ndarray:
+        with self.camera.GrabOne(2000) as result:
+            if result.GrabSucceeded():
+                frame = result.GetArray()
+            else:
+                frame = None
 
-    def __init__(self, url=None, **kwargs):
-        self.url = url
+        frame = self._crop(frame)
+        frame = self._digital_zoom(frame)
 
-    def grab(self):
-        start = time.time()
-        try:
-            req = urllib.request.urlopen(self.url)
-            response = req.read()
-            arr = np.asarray(bytearray(response), dtype=np.uint8)
-            frame = cv2.imdecode(arr, -1)  # 'Load it as it is'
-        except Exception as e:
-            logger.error(f"could not grab frame from {self.url}: {str(e)}")
-            frame = None
-        now = time.time()
-        elapsed = now - start
-        logger.info(f"read image from URL {self.url} into frame in {elapsed}s")
+        frame = cv2.cvtColor(frame, cv2.COLOR_RGB2BGR)
 
         return frame
+
+    def release(self) -> None:
+        self.camera.Close()
+        BaslerUSBFrameGrabber.serial_numbers_in_use.remove(self.serial_number)
+
+    def apply_options(self, options: dict) -> None:
+        self.config["options"] = options
+
+        if self.pixel_format:
+            self.camera.PixelFormat.SetValue(self.pixel_format)
+
+        if self.exposure:
+            self.camera.ExposureTime.SetValue(self.exposure)
+
+    @property
+    def pixel_format(self):
+        return self.config.get("options", {}).get("pixel_format")
+
+    @property
+    def exposure(self):
+        return self.config.get("options", {}).get("exposure")
+
+
+class RealSenseFrameGrabber(FrameGrabber):
+    """Intel RealSense Depth Camera"""
+
+    def __init__(self, config: dict):
+        if rs is None:
+            raise ImportError(
```

Comment:
This is well done.  The only thing I might want to improve is to raise the original `ImportError` in case there's some useful debugging about why it didn't work.  e.g. if you think the library should be installed but it's not working for some reason, this code obscures what's going on.  The `UnavailableModule` class [here](https://github.com/groundlight/python-sdk/blob/main/src/groundlight/optional_imports.py) could help with that, but it's also not a critical improvement.  We could totally run with this for now.

Reply:
Good point. I wonder if we could log a warning instead so that the user doesn't have to handle the exception. In general, I would like to discuss how we use logging in this library. In the current version, I am not sure that we have enough logging, and if the logging is appropriate.